src/main-process/atom-application.coffee

```diff
@@ -648,30 +655,39 @@ class AtomApplication
   #   :devMode - Boolean to control the opened window's dev mode.
   #   :safeMode - Boolean to control the opened window's safe mode.
   openUrl: ({urlToOpen, devMode, safeMode, env}) ->
-    parsedUrl = url.parse(urlToOpen)
+    parsedUrl = url.parse(urlToOpen, true)
     return unless parsedUrl.protocol is "atom:"
 
     pack = @findPackageWithName(parsedUrl.host, devMode)
     if pack?.urlMain
       @openPackageUrlMain(parsedUrl.host, pack.urlMain, urlToOpen, devMode, safeMode, env)
     else
-      @openPackageUriHandler(urlToOpen, devMode, safeMode, env)
-
-  openPackageUriHandler: (url, devMode, safeMode, env) ->
-    resourcePath = @resourcePath
-    if devMode
-      try
-        windowInitializationScript = require.resolve(path.join(@devResourcePath, 'src', 'initialize-application-window'))
-        resourcePath = @devResourcePath
-
-    windowInitializationScript ?= require.resolve('../initialize-application-window')
-    if @lastFocusedWindow?
-      @lastFocusedWindow.sendURIMessage url
+      @openPackageUriHandler(urlToOpen, parsedUrl, devMode, safeMode, env)
+
+  openPackageUriHandler: (url, parsedUrl, devMode, safeMode, env) ->
+    bestWindow = null
+    if parsedUrl.host is 'core'
+      predicate = require('../core-uri-handlers').windowPredicate(parsedUrl)
+      bestWindow = @getLastFocusedWindow (win) ->
+        not win.isSpecWindow() and predicate(win)
+
+    bestWindow ?= @getLastFocusedWindow (win) -> not win.isSpecWindow()
+    if bestWindow?
+      bestWindow.sendURIMessage url
+      bestWindow.focus()
     else
+      resourcePath = @resourcePath
+      if devMode
+        try
+          windowInitializationScript = require.resolve(path.join(@devResourcePath, 'src', 'initialize-application-window'))
+          resourcePath = @devResourcePath
+
+      windowInitializationScript ?= require.resolve('../initialize-application-window')
       windowDimensions = @getDimensionsForNewWindow()
-      @lastFocusedWindow = new AtomWindow(this, @fileRecoveryService, {resourcePath, windowInitializationScript, devMode, safeMode, windowDimensions, env})
-      @lastFocusedWindow.on 'window:loaded', =>
-        @lastFocusedWindow.sendURIMessage url
+      win = new AtomWindow(this, @fileRecoveryService, {resourcePath, windowInitializationScript, devMode, safeMode, windowDimensions, env})
+      @windowStack.addWindow(win)
+      win.on 'window:loaded', ->
+        win.sendURIMessage url
```

Comment:
@BinaryMuse Any way we could use [HistoryManager](https://atom.io/docs/api/v1.21.2/HistoryManager#instance-getProjects) to check if `url` is in a project and load the entire project and open the `url`?

maybe we would want a setting to open individual files or project if found?